src/process_impl/mod.rs

```diff
@@ -0,0 +1,203 @@
+#![allow(
+    missing_docs,
+    missing_debug_implementations,
+    rustdoc::broken_intra_doc_links,
+    unreachable_pub
+)]
+
+use std::ffi::OsStr;
+use std::io;
+use tokio::io::AsyncReadExt;
+use tokio::process;
+
+use super::Error;
+
+use super::{ForwardType, Socket};
+
+pub(crate) mod builder;
+
+mod command;
+pub(crate) use command::Command;
+
+mod child;
+pub(crate) use child::RemoteChild;
+
+#[derive(Debug)]
+pub(crate) struct Session {
+    ctl: tempfile::TempDir,
+    addr: String,
+    terminated: bool,
+    master: std::sync::Mutex<Option<(tokio::process::ChildStdout, tokio::process::ChildStderr)>>,
+}
+
+impl Session {
+    fn ctl_path(&self) -> std::path::PathBuf {
+        self.ctl.path().join("master")
+    }
+
+    pub(crate) async fn check(&self) -> Result<(), Error> {
+        if self.terminated {
+            return Err(Error::Disconnected);
+        }
+
+        let check = process::Command::new("ssh")
+            .arg("-S")
+            .arg(self.ctl_path())
+            .arg("-o")
+            .arg("BatchMode=yes")
+            .arg("-O")
+            .arg("check")
+            .arg(&self.addr)
+            .output()
+            .await
+            .map_err(Error::Ssh)?;
+
+        if let Some(255) = check.status.code() {
+            if let Some(master_error) = self.take_master_error().await {
+                Err(master_error)
+            } else {
+                Err(Error::Disconnected)
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    pub(crate) fn raw_command<S: AsRef<OsStr>>(&self, program: S) -> Command {
+        // XXX: Should we do a self.check() here first?
+
+        // NOTE: we pass -p 9 nine here (the "discard" port) to ensure that ssh does not
+        // succeed in establishing a _new_ connection if the master connection has failed.
+
+        let mut cmd = process::Command::new("ssh");
+        cmd.arg("-S")
+            .arg(self.ctl_path())
+            .arg("-T")
+            .arg("-o")
+            .arg("BatchMode=yes")
+            .arg("-p")
+            .arg("9")
+            .arg(&self.addr)
+            .arg("--")
+            .arg(program);
+
+        Command::new(cmd)
+    }
+
+    pub(crate) async fn request_port_forward(
+        &self,
+        forward_type: ForwardType,
+        listen_socket: &Socket<'_>,
+        connect_socket: &Socket<'_>,
+    ) -> Result<(), Error> {
+        let flag = match forward_type {
+            ForwardType::Local => "-L",
+            ForwardType::Remote => "-R",
+        };
+
+        process::Command::new("ssh")
```

Comment:
We should check the exist code for this command. If it's non-zero, the forwarding didn't work, and the exit status might tell us whether that's because the config doesn't allow them, or because the master has died. And if the latter is the case, we need to do the `take_master_error` dance.

Reply:
Done.